pkg/vm/engine/tae/tables/updates/delchain.go

```diff
@@ -76,29 +76,17 @@ func (chain *DeleteChain) StringLocked() string {
 func (chain *DeleteChain) GetController() *MVCCHandle { return chain.mvcc }
 
 func (chain *DeleteChain) IsDeleted(row uint32, ts types.TS, rwlocker *sync.RWMutex) (deleted bool, err error) {
-	chain.LoopChain(
-		func(vn txnif.MVCCNode) (goNext bool) {
-			n := vn.(*DeleteNode)
-			if n.GetStartTS().Greater(ts) {
-				return true
-			}
-			if n.HasOverlapLocked(row, row) {
-				needWait, txnToWait := n.NeedWaitCommitting(ts)
-				if needWait {
-					rwlocker.RUnlock()
-					txnToWait.GetTxnState(true)
-					rwlocker.RLock()
-				}
-				if n.IsVisible(ts) {
-					deleted = true
-				}
-			}
-			if n.IsMerged() || deleted || err != nil {
-				return false
-			}
-			return true
-		})
-	return
+	deleteNode := chain.GetDeleteNodeByRow(row)
+	if deleteNode == nil {
+		return false, nil
+	}
+	needWait, txn := deleteNode.NeedWaitCommitting(ts)
+	if needWait {
+		rwlocker.RUnlock()
```

Comment:
I never really understand the locking mechanism of TAE -- for example here, why this lock does not belong to txn?   Or a txn manager?    Where did the lock come from?   Everyone dealing with txn repects this lock?

Reply:
The  `rwlocker ` is just a latch , which likes a `mutex`, to prevent `DeleteChain` being accessed concurrently .  It 's not a **lock** for transaction, in TAE , there're no transaction locks .

Reply:
Here I give an example:
```go
type Payload struct {
    value any
    txn   *Transaction
    next  *Payload
}

type Resources struct {
    *sync.RWMutex
    payload *Payload
}

// Current implementation
func getValuesLocked(timestamp int, resources *Resources) any {
    payload := Resources.payload
    for payload != nil {
        if txn.IsVisible(timestamp) {
            if txn.NeedWaitCommitted() {
                resources.RUnlock()
                // wait committed|rollbacked
                resources.RLock()
            }
            // collect value
        }
        payload = payload.next
    }
    return
}
func getValues(timestamp int, resources *Resources) any {
    resources.RLock()
    defer resources.RUnlock()
    return getValuesLocked(timestamp, resources)
}

// Another implementation
func getValuesLocked(timestamp int, resources *Resources) (result any, pendings []*Payload) {
    payload := resources.payload
    for payload != nil {
        if txn.IsVisible(timestamp) {
            if txn.NeedWaitCommitted() {
                pendings = append(pendings, payload)
            } else {
                // collect value
            }
        }
        payload = payload.next
    }
    return
}

func getValues(timestamp int, resources *Resources) any {
    resources.RLock()
    result, pendings := getValuesLocked(timestamp, resources)
    resources.RUnlock()
    // 1. wait pending transactions
    // 2. process pending transactions
    return
}

func main() {
    var timestamp int
    var resources *Resources
    // ...
    getValues(timestamp, resources)
}
```
This current implementation is a bit strange. There are some prerequisite constraints on the data when using it. In fact, it can also be written in another way. In comparison, the other way of implementation is much easier to understand. 

When dealing with waiting during traversal, there are the following considerations: 
1. Our transactions are basically group commits, so waiting for one is basically waiting for a batch
2. Collect the results during the traversal process. For those that require order, one traversal can be reduced. Once there is a transaction to wait for, it will then traverse again after waiting for these transactions.

In fact, in some cases where we do not need to guarantee the order of data, we use another implementation